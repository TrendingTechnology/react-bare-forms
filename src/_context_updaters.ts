// Context Updaters
// Methods to update & track the state of all inputs, validators & general form metadata
//
// For example:
// const context = {
//     // other props...
//     metadata: {
//          inputs: {
//              email: {
//                  isValid: false,
//                  messages: ['Cannot be empty'],
//                  value: '',
//                  isTouched: false,
//              },
//              username: {
//                  isValid: true,
//                  messages: [],
//                  value: 'joebloggs',
//                  isTouched: true,
//              }
//          }
//     }
// };
import {IValidation} from "./validators";


export const updateValidationMetadata = (context: any, update: any) =>
    (fieldName: string, fieldValue: any, validation: IValidation): void => {
        if(typeof context.metadata === "undefined") {
            return;
        }
        if(!(fieldName in context.metadata)) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
            // Should update context.metadata value
        } else if(context.metadata[fieldName].value != fieldValue) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
        }

    };


