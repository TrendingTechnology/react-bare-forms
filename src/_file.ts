import {default as React} from "react";
import {getFileFromRef} from "./uncrontrolled";

// File Helper Functions


/** @internal **/
export interface IFile {
    readonly name: string;
    readonly lastModified: string;
    readonly lastModifiedDate: string;
    readonly size: string;
    readonly type: string;
}

/** @internal */
export function createFileObject(ref: React.RefObject<HTMLFormElement>): IFile {
    try {
        let {name, lastModified, lastModifiedDate, size, type} = getFileFromRef(ref);
        return {
            name,
            lastModified,
            lastModifiedDate,
            size,
            type,
        };
    } catch (err) {
          return null;
    }
}
