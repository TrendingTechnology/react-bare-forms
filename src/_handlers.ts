import * as React from "react";

/** @internal */
type TypeHandler = (e: React.ChangeEvent<any>, name: string) => void;

/** @internal */
export function updateStateFromPassedInContext(parentState: any, setParentState: Function): TypeHandler {
    return (e: React.ChangeEvent<any>, name: string) => {
        setParentState({
            [name]: e.target.value,
        });
    };
}
