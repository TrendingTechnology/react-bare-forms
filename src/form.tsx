import {default as React, useEffect, useState} from "react";
import {updateStateFromPassedInContext} from "./_handlers";
import {IValidation} from "./validators";
import {updateRadioGroupMetadata, updateValidationMetadata} from "./_context_updaters";
import {IRadioField} from "./form-elements";


export interface IRadioGroupChildren {
    name: string;
    isChecked: boolean;
    disabled: boolean; // TODO
}

/** @internal */
export interface _IFormMetadata {
    messages: Array<string>;
    isValid: boolean;
    isTouched: boolean;
    value: any;
}

/** @internal */
export type TypeMetadata = { [k: string]: _IFormMetadata};
export type TypeRadioGroup = {[k: string]: IRadioGroupChildren};

export interface IMetdadata {
    fieldGroups: TypeRadioGroup;
    inputs: TypeMetadata;
}

/**
 * @interface **IForm** Exported Form interface available to the caller. Contains all the properties required by
 * the Form *RBF* Form's component.
 */
export interface IForm extends React.FormHTMLAttributes<HTMLFormElement> {
    /** The passed in state from the parent component */
    state: any;
    /** If the parent component is a class component then the context must contain the parent's **this** keyword. */
    context?: any;
    /** If the form's state is not directly contained at parent's state root object, then formKey needs to represent the key. */
    formKey?: string;
    /** Default is false. If set to true then the form state will be displayed just below the form component in `code` tags. */
    debug?: boolean;
    /** Default is True. If set to false, the  validators will not be called until the form is submitted. */
    dynamic?: boolean;
    /** The passed through props to the Form component */
    children?: any;
    /** Default set to false. If set to true then it will now use bootstrap styling or any extra elements. */
    bare?: boolean;
}

/** @internal */
export interface IFormContext {
    bare?: boolean;
    debug?: boolean;
    dynamic?: boolean;
    formKey?: string;
    metadata: TypeMetadata;
    state: any;
    updateParentState?: (e: React.ChangeEvent<any>, name: string) => void;
    updateFieldValidation?: (fieldName: string, fieldValue: any, validation: IValidation) => void;
    updateFormGroupMetadata?: (fieldGroupKey: string, radioProps: Array<{ props: IRadioField}>) => void;
}


const providerContext: IFormContext = {
    bare: false,
    state: {},
    formKey: null,
    debug: false,
    dynamic: true,
    metadata: { inputs: ({} as any), fieldGroups: ({} as any)},
};

/** @internal */
export const FormContext = React.createContext(providerContext);
/** @internal */
export const FormProvider = FormContext.Provider;
/**
 * @var Access the form state within a React Conext Api Consumer scope
 * @example For example:
 * ```
 * <Form state={this.state} context={this}>
 *      <FormConsumer>
 *      {({state}) => {
 *          return <div>{state} is equal {myState}</div>;
 *      }}
 *    </FormConsumer>
 * </Form>
 * ```
 * */
export const FormConsumer = FormContext.Consumer;


export const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
};

/**
 *
 * @param props
 * @constructor
 */
export const Submit = (props: any) => {
    return <button type="submit">Submit</button>;
};


/**
 * The main Form component
 * @param props
 *
 * ```
 * // Minimal setup for a RBF's Form component
 *
 *  const myState = {
 *      username: '',
 *  }
 *
 *  <Form state={myState}></Form>
 * ```
 * @constructor
 */
export const Form = (props: IForm) => {
    const [parentState, setParentState] = useState(props.state);

    // If the parent component is a class component, then the state needs to be updated from the parent context
    if(props.context) {
        useEffect(() => {
            props.context.setState({
                ...parentState,
            });
        }, [parentState]);
    }


    const [context, updateContext] = useState(providerContext);

    const _providerContext: IFormContext = {
        bare: props.bare || context.bare,
        state: props.state,
        formKey: props.formKey,
        debug: props.debug || context.debug,
        dynamic: props.dynamic || context.dynamic,
        metadata: context.metadata,
        updateParentState: updateStateFromPassedInContext(parentState, setParentState),
        updateFieldValidation: updateValidationMetadata(context, updateContext),
        updateFormGroupMetadata: updateRadioGroupMetadata(context, updateContext),
    };

    return (
        <FormProvider value={_providerContext}>
            <form onSubmit={handleSubmit} {...props}>{props.children}</form>
        </FormProvider>
    );
};
