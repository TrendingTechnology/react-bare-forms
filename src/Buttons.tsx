import {ReactElement, default as React, Consumer} from "react";
import {FormConsumer, IFormContext} from "./form";

export interface IButton<T> {
    create(): (props: ISubmitButtonProps) => ReactElement<T>;
}


abstract class Button<T> implements IButton<T> {

    public abstract create(): (props: T) => ReactElement<T>;

}

export interface ISubmitButtonProps {
    text?: string;
    disabled?: false;
    className?: string;
    children?: any;
}

/**
 * SubmitButton is the main component for submitting forms.
 *
 */
class _SubmitButton implements Button<ISubmitButtonProps> {

    create() {
       return (props: ISubmitButtonProps) => {
           const { className = "btn btn-success btn-lg", disabled = true} = props;
           return (
               <FormConsumer>
                   {(context: IFormContext) => {
                       let isDisabled;
                       if(context && context.metadata && Object.keys(context.metadata.inputs).length > 0) {
                           const inputItem = context.metadata.inputs;
                           Object.keys(inputItem).map((key: any) => {
                               if(!inputItem[key].isValid) {
                                   isDisabled = true;
                               }
                           });
                       }
                       return (<button
                           disabled={isDisabled}
                           className={className}
                       >{props.children}</button>);
                   }}
               </FormConsumer>
           );
       }
    }
}

export const SubmitButton = new _SubmitButton().create();
