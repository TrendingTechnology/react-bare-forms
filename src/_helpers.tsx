import * as React from "react";
import {handleChange} from "./handlers";
import {FieldTypes} from "./form-elements";
import {IFormContext} from "./form";
import {IValidation} from "./validators";

/**
 *
 * @param fieldType
 * @param name
 * @param context
 */
export const selectTextField = (fieldType: FieldTypes, name: any, context: IFormContext) => {
    switch(fieldType) {
        case "text": {
            return <input type="text" className="form-control" onChange={handleChange(name, context)} />;
        }
        case "email": {
            return <input type="email" className="form-control" onChange={handleChange(name, context)} />;
        }
        case "password": {
            // TODO confirm password
            return <input type="password" className="form-control" onChange={handleChange(name, context)} />;
        }
        default: {
            return <input type="text" className="form-control" onChange={handleChange(name, context)} />;
        }
    }
};

/**
 *
 * @param validationResult
 * @param context
 */
export const shouldShowValidation = (validationResult: IValidation, context: IFormContext) => {
    return (!validationResult.isValid && context.dynamic) ||
        (!validationResult.isValid && context.isSubmitted);
};


/**
 *
 * @param update
 * @param current
 */
export function setFormData(update: Function, current: IFormContext) {
    return (name: any, value: any) => {
        update({
            ...current,
            state: {
                ...current.state,
                [current.formKey]: {
                    ...current.state[current.formKey],
                    [name]: value,
                }
            }
        });
    };
}
