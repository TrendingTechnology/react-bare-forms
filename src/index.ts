import {Form, FormConsumer} from "./form";
import {
    CheckBoxField,
    EmailField, FileField,
    PasswordField, RadioField, RadioGroup,
    SelectField,
    SubmitButton,
    TextAreaField,
    TextInputField
} from "./elements";
import {areFieldsEqual, isChecked, isEmailValid, isFieldEmpty, isFile, isRadioChecked} from "./validators";
import {createFileRef, getFileFromRef} from "./uncrontrolled";




export {
    // Uncontrolled
    createFileRef,
    getFileFromRef,
    // Validators
    areFieldsEqual,
    isChecked,
    isEmailValid,
    isFieldEmpty,
    isFile,
    isRadioChecked,
    // Form Components
    FormConsumer,
    Form,
    SubmitButton,
    TextInputField,
    EmailField,
    PasswordField,
    TextAreaField,
    CheckBoxField,
    SelectField,
    FileField,
    RadioField,
    RadioGroup,
}
