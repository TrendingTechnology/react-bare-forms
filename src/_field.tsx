import {FormContext, IFormContext} from "./form";
import {FormElementValidators, mergeDefaultCssWithProps} from "./_helpers";
import {ChangeEvent, default as React, ReactElement, useContext, useEffect} from "react";
import {
    IFileField,
    RadioGroupContext,
    TypeSelectCssSizeName
} from "./elements";
import {shouldUpdateRadioGroupContext} from "./_context_updaters";

interface IFieldClass<T> {
    create: (context: IFormContext) => ReactElement<T>;
    type: string;
    props: T;
}

/** @internal */
function _genericFormGroup<T extends any>(props: T, children: any) {
    return (
        <div className="form-group">
            {props.labelText && <label>{props.labelText}</label>}
            {children}
            {props.hint && <small className="form-text text-muted">{props.hint}</small>}
        </div>
    );
}

/** @internal */
abstract class _field<PropsType extends any> {

    type: string;

    props: PropsType;

    protected constructor(props: PropsType, type?: string) {
        this.type = type;
        this.props = props;
    }

    public createField(fieldCallback: Function) {
        const context: IFormContext = useContext(FormContext);
        const _validate = this.props.validators ?
            <FormElementValidators
                validators={this.props.validators}
                name={this.props.name}
                value={this.props.value}
                type="inputs"
            /> :
            null;
        if(context.bare) {
            return (<>{fieldCallback(context)}{_validate}</>);
        } else {
            return(<>{this.formGroup(fieldCallback(context))}{ _validate}</>);
        }
    }

    public abstract formGroup(children: any): ReactElement;

    public abstract getField(): (context: IFormContext) => ReactElement;

    static mergeDefaultCssWithProps(defaultValue: string, cssProps: any, bare: boolean): string {
        let cssStr = "";
        if(!bare) {
            cssStr += `${defaultValue} `;
        }
        if (cssProps) {
            cssStr += `${cssProps}`;
        }
        return cssStr;
    }

    public overrideEvent(e: any, value: any) {
        return {
            ...e,
            target: {
                ...e.target,
                value: !value,
            }
        }
    }

}

/** @internal */
export class InputField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        return (context: IFormContext) => {
            return <input
                type={this.type}
                value={context.state[this.props.name] || ""}
                onChange={(e) => context.updateParentState(e, this.props.name)}
                name={this.props.name}
                className={_field.mergeDefaultCssWithProps("form-control", this.props.className, context.bare)}
            />;
        }
    }

}

/** @internal */
export class CheckBoxField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return (
            <div className="form-group form-check">
                {children}
                {this.props.labelText && <label className="form-check-label">{this.props.labelText}</label>}
                {this.props.hint && <small className="form-text text-muted">{this.props.hint}</small>}
            </div>
        );
    }

    public getField() {
        return (context: IFormContext) => {
            return <input
                type={this.type}
                checked={context.state[this.props.name] || false}
                onChange={(e) => context.updateParentState(this.overrideEvent(e, context.state[this.props.name]), this.props.name)}
                name={this.props.name}
                className={_field.mergeDefaultCssWithProps("form-check-input", this.props.className, context.bare)}
            />;
        }
    }
}

/** @internal */
export class TextAreaField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        const {rows = 5} = this.props;
        return (context: IFormContext) => {
            return (
                <textarea
                    className={mergeDefaultCssWithProps("form-control", this.props.className, context.bare)}
                    rows={rows}
                    value={context.state[this.props.name] || ""}
                    onChange={(e) => context.updateParentState(e, this.props.name)}
                    name={this.props.name}
                />
            );
        }
    }
}

/** @internal */
export class RadioField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return (
            <div className="form-group form-check">
                {children}
                {this.props.labelText && <label className="form-check-label">{this.props.labelText}</label>}
                {this.props.hint && <small className="form-text text-muted">{this.props.hint}</small>}
            </div>
        );
    }

    public getField() {
        const radioContext: any = useContext(RadioGroupContext); // TODO - any
        return (context: IFormContext) => {
        const updateContexts = (e: React.ChangeEvent<any>, context: IFormContext) => {
            context.updateRadioGroupStateFromPassedInContext(
                this.overrideEvent(e, context.state[this.props.name]),
                this.props.name,
                radioGroup,
            );
            if(shouldUpdateRadioGroupContext(radioContext.children, context, radioContext.parent.name)) {
                useEffect(() => {
                    context.updateRadioGroupMetadata(radioContext.parent.name, radioContext.children);
                }, [radioContext]);

            }
        };
        const radioGroup = context.metadata.fieldGroups[radioContext.parent.name];
        return <input
            type={this.type}
            checked={context.state[this.props.name] || false}
            name={this.props.name}
            className={_field.mergeDefaultCssWithProps("form-check-input", this.props.className, context.bare)}
        />

        }
    }

}

/** @internal */
export class SelectField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        const {options = [], size = "default"} = this.props;
        return (context: IFormContext) => {

            return (
                <select
                    onChange={(e) => context.updateParentState(e, this.props.name)}
                    name={this.props.name}
                    className={_field.mergeDefaultCssWithProps(this.getSelectCssName(this.props.size), this.props.className, context.bare)}
                >
                    {options.map((optVal: string, i: number) => {
                        return <option value={optVal} key={i}>{optVal}</option>
                    })}
                </select>
            );
        }
    }

    private getSelectCssName(name: TypeSelectCssSizeName) {
        if(name === "default") {
            return "form-control"
        } else {
            return `form-control form-control-${name}`;
        }
    }
}

/** @internal */
export class FileField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        return (context: IFormContext) =>
            (<input
                ref={this.props.ref}
                type="file"
                className={_field.mergeDefaultCssWithProps("form-control-file", this.props.className, context.bare)}
            />);
    }
}
