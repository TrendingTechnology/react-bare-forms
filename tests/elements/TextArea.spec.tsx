import * as React from "react";
import renderer from "react-test-renderer";
import {Form, TextAreaField,isFieldEmpty} from "../../src";
import { MockComponent } from "../form.spec";



describe("#<TextAreaField />", () => {
    it("should render an input field with the bootstrap styled tags", () => {
        let state = {
            about_text: "joe",
        };


        let component = renderer.create(
            <MockComponent state={state}>
                <TextAreaField name="about_text" value={state.about_text} />
            </MockComponent>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a single input tag with no bootstrap styles", () => {
        let state = {
            about_text: "",
        };

        let component = renderer.create(
            <MockComponent state={state}>
                <TextAreaField name="about_text" value={state.about_text} />
            </MockComponent>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });


    it("it should pass in props", () => {
        let state = {
            about_text: "joebloggs"
        };

        let testFormRederer: any = renderer.create(
            <MockComponent state={state}>
                <TextAreaField name="about_text" value={state.about_text} />
            </MockComponent>
        );

        const testFormInstance = testFormRederer.root;
        expect(testFormInstance.props.state.about_text).toEqual(state.about_text);

    });

    it("it should display validation errors", () => {
        let state = {
            about_text: "a",
        };

        let component = renderer.create(
            <MockComponent state={state}>
                <TextAreaField
                    rows={20}
                    name="about_text"
                    value={state.about_text}
                    validators={[isFieldEmpty(5)]}
                />
            </MockComponent>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("it should pass validation props", () => {
        let state = {
            about_text: "a",
        };

        let testFormRederer: any = renderer.create(
            <MockComponent state={state}>
                <TextAreaField
                    rows={10}
                    name="about_text"
                    value={state.about_text}
                    validators={[isFieldEmpty(5)]}
                />
            </MockComponent>
        );

        const testFormInstance = testFormRederer.root;
        // expect(testFormInstance.props.value).toEqual(state.about_text);
        expect(testFormInstance);

    });
});

