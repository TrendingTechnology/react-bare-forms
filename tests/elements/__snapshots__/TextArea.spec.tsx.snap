// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#<TextAreaField /> it should display validation errors 1`] = `
<form
  bare="false"
  context={[Function]}
  onSubmit={[Function]}
  state={
    Object {
      "about_text": "a",
    }
  }
>
  <textarea
    className=""
    name="about_text"
    onChange={[Function]}
    rows={20}
    value="a"
  />
</form>
`;

exports[`#<TextAreaField /> should render a single input tag with no bootstrap styles 1`] = `
<form
  bare="false"
  context={[Function]}
  onSubmit={[Function]}
  state={
    Object {
      "about_text": "",
    }
  }
>
  <textarea
    className=""
    name="about_text"
    onChange={[Function]}
    rows={5}
    value=""
  />
</form>
`;

exports[`#<TextAreaField /> should render an input field with the bootstrap styled tags 1`] = `
<form
  bare="false"
  context={[Function]}
  onSubmit={[Function]}
  state={
    Object {
      "about_text": "joe",
    }
  }
>
  <textarea
    className=""
    name="about_text"
    onChange={[Function]}
    rows={5}
    value="joe"
  />
</form>
`;

exports[`<Form /> should render a form element with bootstrap extra tags 1`] = `
<form
  bare={false}
  context={[Function]}
  onSubmit={[Function]}
  state={
    Object {
      "username": "joebloggs",
    }
  }
/>
`;

exports[`<Form /> should render a form element with no extra tags 1`] = `
<form
  bare="false"
  context={[Function]}
  onSubmit={[Function]}
  state={
    Object {
      "username": "joebloggs",
    }
  }
/>
`;
